sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,481 @@
+use omicron_common::disk::{DatasetKind, DatasetsConfig};
+use omicron_common::update::ArtifactHash;
+use repo_depot_api::*;
+use sha2::{Digest, Sha256};
+use sled_storage::manager::StorageHandle;
+use slog::{info, Logger};
+use tokio::fs::File;
+use tokio::io::AsyncWriteExt;
+
+#[derive(Clone)]
+pub(crate) struct ArtifactStore<T: StorageBackend = StorageHandle> {
+    log: Logger,
+    storage: T,
+}
+
+impl<T: StorageBackend> ArtifactStore<T> {
+    pub(crate) fn new(log: &Logger, storage: T) -> ArtifactStore<T> {
+        ArtifactStore {
+            log: log.new(slog::o!("component" => "ArtifactStore")),
+            storage,
+        }
+    }
+}
+
+impl ArtifactStore {
+    pub(crate) fn start(
+        self,
+        sled_address: SocketAddrV6,
+        dropshot_config: &ConfigDropshot,
+    ) -> Result<
+        dropshot::HttpServer<ArtifactStore<StorageHandle>>,
+        Box<dyn std::error::Error + Send + Sync>,
+    > {
+        let mut depot_address = sled_address;
+        depot_address.set_port(REPO_DEPOT_PORT);
+
+        let log = self.log.new(o!("component" => "dropshot (Repo Depot)"));
+        Ok(HttpServerStarter::new(
+            &ConfigDropshot {
+                bind_address: depot_address.into(),
+                ..dropshot_config.clone()
+            },
+            repo_depot_api_mod::api_description::<RepoDepotImpl>()
+                .expect("registered entrypoints"),
+            self,
+            &log,
+        )?
+        .start())
+    }
+}
+
+impl<T: StorageBackend> ArtifactStore<T> {
+    /// GET operation (served by Repo Depot API)
+    pub(crate) async fn get(
+        &self,
+        sha256: ArtifactHash,
+    ) -> Result<File, Error> {
+        let sha256 = sha256.to_string();
+        for dataset in self.datasets().await? {
+            let path = dataset.join(&sha256);
+            match File::open(&path).await {
+                Ok(file) => {
+                    info!(
+                        &self.log,
+                        "Retrieved artifact";
+                        "sha256" => &sha256,
+                        "path" => path.as_str()
+                    );
+                    return Ok(file);
+                }
+                Err(err) if err.kind() == ErrorKind::NotFound => {}
+                Err(err) => {
+                    return Err(Error::File { verb: "open", path, err });
+                }
+            }
+        }
+        Err(Error::NotFound { sha256 })
+    }
+
+    /// PUT operation (served by Sled Agent API)
+    pub(crate) async fn put(
+        &self,
+        sha256: ArtifactHash,
+        stream: impl Stream<Item = Result<impl AsRef<[u8]>, Error>>,
+    ) -> Result<(), Error> {
+        let ds = self.datasets().await?.next().ok_or(Error::NoUpdateDataset)?;
+        let final_path = ds.join(sha256.to_string());
+
+        let (file, temp_path) = tokio::task::spawn_blocking(move || {
+            NamedUtf8TempFile::new_in(&ds).map_err(|err| Error::File {
```

Comment:
So, this is a nitpick based on the implementation of `camino_tempfile` and `tempfile`:

- I think it's a good idea to write to a "temporary file" and then to "persist" it later, once the write is complete
- I think it's nice that, if we bail out and drop the file, we'll try to delete the temporary file

However, I think that there's a case where we'd leak temporary files here:

https://docs.rs/camino-tempfile/latest/camino_tempfile/struct.NamedUtf8TempFile.html#resource-leaking

If, between creation of the temporary file and calling `persist`, the sled agent dies, we will have a "temporary file" on a durable update dataset that will never get deleted.

I think we need to handle this case explicitly - for example, we could create a directory within this dataset for temporary files (which the sled agent could remove on reboot), or have some naming convention which makes these in-progress artifacts easier to find (e.g., maybe their `final_path` plus a `.tmp` suffix or something)

Reply:
The subdirectory option is pretty good, since we can `tokio::fs::remove_dir_all` it when the artifact store is initially spun up.